libvast/src/detail/load_plugin.cpp

```diff
@@ -0,0 +1,109 @@
+  if (auto binary = objectpath(nullptr))
+    result.insert(binary->parent().parent() / "lib" / "vast" / "plugins");
+  else
+    VAST_ERROR("{} failed to get program path", __func__);
+  if (const char* home = std::getenv("HOME"))
+    result.insert(path{home} / ".local" / "lib" / "vast" / "plugins");
+  if (auto dirs = caf::get_if<std::vector<std::string>>( //
+        &cfg, "vast.plugin-dirs"))
+    result.insert(dirs->begin(), dirs->end());
+  return result;
+}
+
+} // namespace
+
+caf::expected<std::pair<path, plugin_ptr>>
+load_plugin(path file, caf::actor_system_config& cfg) {
+  auto& plugins = plugins::get();
+  auto try_load_plugin = [&](path file) -> caf::expected<plugin_ptr> {
+#if VAST_MACOS
+    if (file.extension() == "")
+      file = file.str() + ".dylib";
+#else
+    if (file.extension() == "")
+      file = file.str() + ".so";
+#endif
+    if (!exists(file))
+      return caf::no_error;
+    if (auto plugin = plugin_ptr::make(file.str().c_str(), cfg)) {
+      VAST_ASSERT(*plugin);
+      auto has_same_name = [name = (*plugin)->name()](const auto& other) {
+        return !std::strcmp(name, other->name());
+      };
+      if (std::none_of(plugins.begin(), plugins.end(), has_same_name)) {
+        return plugin;
+      } else {
+        return caf::make_error(ec::invalid_configuration,
+                               fmt::format("failed to load plugin {} because "
+                                           "another plugin already uses the "
+                                           "name {}",
+                                           file, (*plugin)->name()));
+      }
+    } else {
+      return std::move(plugin.error());
+    }
+    return ec::logic_error;
+  };
+  auto load_errors = std::vector<caf::error>{};
+  // First, check if the plugin file is specified as an absolute path.
+  if (auto plugin = try_load_plugin(file))
+    return std::pair{file, std::move(*plugin)};
+  else if (plugin.error() != caf::no_error)
+    load_errors.push_back(std::move(plugin.error()));
+  // Second, check if the plugin file is specified relative to the specified
+  // plugin directories.
+  for (const auto& dir : get_plugin_dirs(cfg))
+    if (auto plugin = try_load_plugin(dir / file))
+      return std::pair{dir / file, std::move(*plugin)};
+    else if (plugin.error() != caf::no_error)
+      load_errors.push_back(std::move(plugin.error()));
+  // We didn't find the plugin, and did not encounter any errors, so the file
+  // just does not exist.
+  if (load_errors.empty())
+    return caf::make_error(ec::invalid_configuration,
+                           fmt::format("failed to find plugin {}", file));
+  // We found the file, but encounterd errors trying to load it.
+  return caf::make_error(ec::invalid_configuration,
+                         fmt::format("failed to load plugin; these partial "
+                                     "errors occured:\n - {}",
+                                     fmt::join(load_errors, "\n - ")));
```

Comment:
I tested the error formatting and was able to generate these:
- with version mismatch:
  ```
  failed to load plugin libclash: !! invalid_configuration: "failed to load plugin; these partial errors occured:\n - !! version_error: \"libvast version mismatch in\" \"/home/tobim/t/build/vast/debug/lib64/vast/plugins/libclash.so\" \"2021.01.28-67-g596de1a8c\" \"2021.01.28-93-ge54083d27\""
  ```
- with type id range clash:
  ```
  failed to load plugin libclash: !! invalid_configuration: "failed to load plugin; these partial errors occured:\n - !! system_error: \"encountered type ID block clash in\" \"/home/tobim/t/build/vast/clash-plugin/lib64/vast/plugins/libclash.so\""
  ```
It would be nice if we could remove the `!! invalid_configuration: "failed to load plugin; these partial errors occured:\n - ` part.

Reply:
Should be better now, please check back.